plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -100,42 +100,46 @@ the term is a constant, extract the type tag and check it for equality with the
 Things work the same way for a fully monomorphized polymorphic type, i.e. @(Integer, Bool@) is not
 any different from just 'Integer' unlifting-wise.
 
-However there's no sensible way of unlifting a value of, say, @[a]@ where @a@ in not a built-in
-type. So let's say we instantiated @a@ to an @Opaque val rep@ like we do for polymorphic functions
-that don't deal with polymorphic built-in types (e.g. @id@, @ifThenElse@ etc). That would mean we'd
-need to write a function from a @[a]@ for some arbitrary built-in @a@ to @[Opaque val a]@. Which
-is really easy to do: it's just @map makeKnown@. But the problem is, unlifting is supposed to be
-cheap and that @map@ is O(n), so for example 'MkCons' would become an O(n) operation making
+When it comes to polymorphic types, unlifting a @val@ into an @Opaque val a@ is trivial - it's just
+@makeKnown@. However, there's no sensible way of unlifting a @val@ into, say, @[Opaque val a]@
```

Comment:
`makeKnown` is an implementation detail and doesn't belong to this file. The file doesn't have any mentions of `readKnown`, `makeKnown`, `ReadKnownIn`, `MakeKnownIn`, `KnownTypeAst` etc. Plus, all of that gets optimized away and doesn't appear in the generated Core for builtins, so those are just some intermediate auxiliary abstractions and we shouldn't pollute the high-level description of the built-in functions machinery with them.

And the fact that it's just `makeKnown` doesn't tell the reader much anyway: we are talking about theoretic limitations here, we could've implemented the slow version and unlift a `val` as a `list` and it also would be `makeKnown`.

... except not `makeKnown`, which is _normally_ for lifting, but `readKnown`, which is _normally_ for unlifting. It's just that if go with unlifting to `[Opaque val a]`, we have to do _lifting during unlifting_. We don't do that anywhere in our instances, but for that `[Opaque val a]` we'd have to. The opposite is also true: we have to do unlifting during lifting of an `[Opaque val a]`.

We also never lift to or unlift from a `val`, it's always `Opaque val a`.

The original wording of these two lines was more precise: it was saying that we couldn't do something for any `[a]` when `a` is not a builtin. In your version it's not clear if it's specific to `Opaque` or not (for example, if `SomeConstant` is OK unlike `Opaque`).

I'm not trying to overwhelm you with criticism or something, just clarifying. I appreciate you rewording the original version, clearly it was confusing (especially the part where my usages of "unlifting" are wildly inconsistent), but we'll need to come up with a more fitting rewording.